File: services/collectors/official_marvel_rivals/article_parser.py
from __future__ import annotations

import logging
from dataclasses import dataclass
from urllib.parse import urldefrag, urljoin, urlparse

from bs4 import BeautifulSoup, Tag

from services.collectors.official_marvel_rivals.news_fetcher import NewsArticleSummary, fetch_html
from services.date_utils import ParsedArticleDate, parse_article_date
from services.media_parser import extract_article_media


logger = logging.getLogger(__name__)

MAX_ARTICLE_TEXT_LENGTH = 12000


@dataclass(frozen=True)
class ParsedArticle:
    title: str
    canonical_url: str
    article_url: str
    raw_date: str | None
    date_info: ParsedArticleDate | None
    body_text: str
    raw_excerpt: str | None
    media_url: str | None
    media_urls: tuple[str, ...]
    media_type: str


class ArticleParser:
    def __init__(self, article_timezone: str) -> None:
        self.article_timezone = article_timezone

    async def fetch_and_parse(self, summary: NewsArticleSummary) -> ParsedArticle:
        html = await fetch_html(summary.article_url)
        if html is None:
            logger.warning("Using news-list fallback data for article %s", summary.article_url)
            return self.from_summary(summary)

        try:
            return self.parse(
                html=html,
                summary=summary,
            )
        except Exception:
            logger.exception("Failed to parse article page %s", summary.article_url)
            return self.from_summary(summary)

    def parse(self, *, html: str, summary: NewsArticleSummary) -> ParsedArticle:
        soup = BeautifulSoup(html, "html.parser")
        body_container = _select_body_container(soup)
        raw_date = _extract_article_date(soup) or summary.raw_date
        date_info = parse_article_date(raw_date, self.article_timezone)
        canonical_url = _canonical_article_url(soup, summary.article_url)
        title = _extract_title(soup) or summary.title
        body_text = _extract_body_text(body_container or soup)
        if not body_text and summary.raw_excerpt:
            body_text = summary.raw_excerpt

        body_text = _truncate_text(body_text, MAX_ARTICLE_TEXT_LENGTH)
        media = extract_article_media(
            soup,
            summary.article_url,
            body_container=body_container,
            fallback_cover_url=summary.cover_image_url,
        )

        return ParsedArticle(
            title=title,
            canonical_url=canonical_url,
            article_url=summary.article_url,
            raw_date=raw_date,
            date_info=date_info,
            body_text=body_text,
            raw_excerpt=summary.raw_excerpt,
            media_url=media.media_url,
            media_urls=media.media_urls,
            media_type=media.media_type,
        )

    def from_summary(self, summary: NewsArticleSummary) -> ParsedArticle:
        date_info = parse_article_date(summary.raw_date, self.article_timezone)
        return ParsedArticle(
            title=summary.title,
            canonical_url=summary.canonical_url,
            article_url=summary.article_url,
            raw_date=summary.raw_date,
            date_info=date_info,
            body_text=summary.raw_excerpt or "",
            raw_excerpt=summary.raw_excerpt,
            media_url=summary.cover_image_url,
            media_urls=(summary.cover_image_url,) if summary.cover_image_url else (),
            media_type="photo" if summary.cover_image_url else "none",
        )


def _select_body_container(soup: BeautifulSoup) -> Tag | None:
    for selector in (
        ".artText",
        ".article-content",
        ".news-content",
        ".content",
        "article",
        "main",
    ):
        container = soup.select_one(selector)
        if container is not None:
            return container

    return None


def _extract_title(soup: BeautifulSoup) -> str | None:
    title_node = soup.select_one("h1.artTitle, h1")
    if title_node is not None:
        title = _clean_text(title_node.get_text(" ", strip=True))
        if title:
            return title

    for selector in ('meta[property="og:title"]', 'meta[name="twitter:title"]'):
        title = _meta_content(soup, selector)
        if title and not _looks_generic_title(title):
            return _clean_title(title)

    if soup.title and soup.title.string:
        return _clean_title(soup.title.string)

    return None


def _extract_article_date(soup: BeautifulSoup) -> str | None:
    time_node = soup.select_one("time[datetime]")
    if time_node is not None:
        value = time_node.get("datetime")
        if value:
            return _clean_text(str(value))

    for selector in (
        ".date",
        ".time",
        ".publish-time",
        ".pubdate",
        'meta[property="article:published_time"]',
        'meta[name="article:published_time"]',
        'meta[name="publishdate"]',
        'meta[name="pubdate"]',
        'meta[name="date"]',
    ):
        if selector.startswith("meta"):
            value = _meta_content(soup, selector)
        else:
            node = soup.select_one(selector)
            value = node.get_text(" ", strip=True) if node else None
        if value:
            return _clean_text(value)

    return None


def _canonical_article_url(soup: BeautifulSoup, article_url: str) -> str:
    link = soup.select_one('link[rel="canonical"][href]')
    if link is not None:
        canonical = _normalize_article_url(str(link.get("href")), article_url)
        if canonical:
            return canonical

    for selector in ('meta[property="og:url"]', 'meta[name="og:url"]'):
        og_url = _meta_content(soup, selector)
        canonical = _normalize_article_url(og_url, article_url) if og_url else None
        if canonical and not _looks_like_homepage(canonical):
            return canonical

    normalized = _normalize_article_url(article_url, article_url)
    return normalized or article_url


def _extract_body_text(container: Tag) -> str:
    for node in container.select("script, style, noscript, iframe, nav, footer, header, .footer, .header-nav"):
        node.decompose()

    blocks: list[str] = []
    seen: set[str] = set()
    for node in container.find_all(["h2", "h3", "h4", "p", "li"]):
        text = _clean_text(node.get_text(" ", strip=True))
        if not text or _is_unrelated_text(text) or text in seen:
            continue

        seen.add(text)
        blocks.append(text)

    if blocks:
        return "\n".join(blocks)

    text = _clean_text(container.get_text(" ", strip=True))
    return "" if _is_unrelated_text(text) else text


def _is_unrelated_text(text: str) -> bool:
    normalized = text.lower().replace(" ", "")
    social_footer = "discord|x|facebook|instagram|tiktok|youtube|twitch"
    if normalized == social_footer:
        return True

    blocked = {
        "download",
        "log in",
        "my account log out",
        "privacy policy",
        "term of use",
        "support",
        "comming soon",
    }
    return text.strip().lower() in blocked


def _meta_content(soup: BeautifulSoup, selector: str) -> str | None:
    tag = soup.select_one(selector)
    if tag is None:
        return None

    content = tag.get("content")
    return _clean_text(str(content)) if content else None


def _normalize_article_url(value: str, base_url: str) -> str | None:
    absolute_url = urljoin(base_url, value.strip())
    absolute_url, _fragment = urldefrag(absolute_url)
    parsed = urlparse(absolute_url)
    if parsed.scheme not in {"http", "https"} or not parsed.netloc:
        return None

    return absolute_url


def _looks_like_homepage(url: str) -> bool:
    path = urlparse(url).path.rstrip("/")
    return path in {"", "/index.html"}


def _looks_generic_title(value: str) -> bool:
    return "super hero team-based pvp shooter" in value.lower()


def _clean_title(value: str) -> str:
    title = _clean_text(value)
    for separator in ("_Marvel Rivals", " | Marvel Rivals", " - Marvel Rivals"):
        if separator in title:
            title = title.split(separator, 1)[0]
    return title.strip()


def _clean_text(value: object) -> str:
    return " ".join(str(value).replace("\xa0", " ").split())


def _truncate_text(value: str, max_length: int) -> str:
    if len(value) <= max_length:
        return value

    truncated = value[:max_length].rsplit("\n", 1)[0].strip()
    return truncated or value[:max_length].strip()

File: services/collectors/official_marvel_rivals/article_parser.js
import { parseArticleDate } from "../../date_utils.js";
import { getText, loadHtml, selectOne } from "../../html.js";
import { getLogger } from "../../logger.js";
import { extractArticleMedia } from "../../media_parser.js";
import { charLength, collapseWhitespace, rsplitOnce, sliceChars } from "../../pyutils.js";
import { urldefrag, urljoin, urlsplit } from "../../urlutils.js";
import { fetchHtml } from "./news_fetcher.js";

const logger = getLogger("services.collectors.official_marvel_rivals.article_parser");

export const MAX_ARTICLE_TEXT_LENGTH = 12000;

/**
 * @typedef {{title: string, canonical_url: string, article_url: string,
 *            raw_date: string|null, date_info: object|null, body_text: string,
 *            raw_excerpt: string|null, media_url: string|null,
 *            media_urls: string[], media_type: string}} ParsedArticle
 */

export class ArticleParser {
  constructor(articleTimezone) {
    this.article_timezone = articleTimezone;
  }

  async fetchAndParse(summary) {
    const html = await fetchHtml(summary.article_url);
    if (html === null) {
      logger.warning(`Using news-list fallback data for article ${summary.article_url}`);
      return this.fromSummary(summary);
    }

    try {
      return this.parse({ html, summary });
    } catch (error) {
      logger.exception(`Failed to parse article page ${summary.article_url}`, error);
      return this.fromSummary(summary);
    }
  }

  parse({ html, summary }) {
    const $ = loadHtml(html);
    const bodyContainer = selectBodyContainer($);
    const rawDate = extractArticleDate($) ?? summary.raw_date;
    const dateInfo = parseArticleDate(rawDate, this.article_timezone);
    const canonicalUrl = canonicalArticleUrl($, summary.article_url);
    const title = extractTitle($) ?? summary.title;
    let bodyText = extractBodyText($, bodyContainer ?? $.root());
    if (!bodyText && summary.raw_excerpt) {
      bodyText = summary.raw_excerpt;
    }

    bodyText = truncateText(bodyText, MAX_ARTICLE_TEXT_LENGTH);
    const media = extractArticleMedia($, summary.article_url, {
      bodyContainer,
      fallbackCoverUrl: summary.cover_image_url,
    });

    return {
      title,
      canonical_url: canonicalUrl,
      article_url: summary.article_url,
      raw_date: rawDate,
      date_info: dateInfo,
      body_text: bodyText,
      raw_excerpt: summary.raw_excerpt,
      media_url: media.media_url,
      media_urls: media.media_urls,
      media_type: media.media_type,
    };
  }

  fromSummary(summary) {
    const dateInfo = parseArticleDate(summary.raw_date, this.article_timezone);
    return {
      title: summary.title,
      canonical_url: summary.canonical_url,
      article_url: summary.article_url,
      raw_date: summary.raw_date,
      date_info: dateInfo,
      body_text: summary.raw_excerpt || "",
      raw_excerpt: summary.raw_excerpt,
      media_url: summary.cover_image_url,
      media_urls: summary.cover_image_url ? [summary.cover_image_url] : [],
      media_type: summary.cover_image_url ? "photo" : "none",
    };
  }
}

function selectBodyContainer($) {
  for (const selector of [".artText", ".article-content", ".news-content", ".content", "article", "main"]) {
    const container = selectOne($, selector);
    if (container !== null) {
      return container;
    }
  }

  return null;
}

function extractTitle($) {
  const titleNode = selectOne($, "h1.artTitle, h1");
  if (titleNode !== null) {
    const title = cleanText(getText(titleNode));
    if (title) {
      return title;
    }
  }

  for (const selector of ['meta[property="og:title"]', 'meta[name="twitter:title"]']) {
    const title = metaContent($, selector);
    if (title && !looksGenericTitle(title)) {
      return cleanTitle(title);
    }
  }

  const documentTitle = selectOne($, "title");
  if (documentTitle !== null && getText(documentTitle)) {
    return cleanTitle(getText(documentTitle));
  }

  return null;
}

function extractArticleDate($) {
  const timeNode = selectOne($, "time[datetime]");
  if (timeNode !== null) {
    const value = timeNode.attr("datetime");
    if (value) {
      return cleanText(String(value));
    }
  }

  for (const selector of [
    ".date",
    ".time",
    ".publish-time",
    ".pubdate",
    'meta[property="article:published_time"]',
    'meta[name="article:published_time"]',
    'meta[name="publishdate"]',
    'meta[name="pubdate"]',
    'meta[name="date"]',
  ]) {
    let value;
    if (selector.startsWith("meta")) {
      value = metaContent($, selector);
    } else {
      const node = selectOne($, selector);
      value = node ? getText(node) : null;
    }
    if (value) {
      return cleanText(value);
    }
  }

  return null;
}

function canonicalArticleUrl($, articleUrl) {
  const link = selectOne($, 'link[rel="canonical"][href]');
  if (link !== null) {
    const canonical = normalizeArticleUrl(String(link.attr("href")), articleUrl);
    if (canonical) {
      return canonical;
    }
  }

  for (const selector of ['meta[property="og:url"]', 'meta[name="og:url"]']) {
    const ogUrl = metaContent($, selector);
    const canonical = ogUrl ? normalizeArticleUrl(ogUrl, articleUrl) : null;
    if (canonical && !looksLikeHomepage(canonical)) {
      return canonical;
    }
  }

  const normalized = normalizeArticleUrl(articleUrl, articleUrl);
  return normalized || articleUrl;
}

function extractBodyText($, container) {
  container.find("script, style, noscript, iframe, nav, footer, header, .footer, .header-nav").remove();

  const blocks = [];
  const seen = new Set();
  container.find("h2, h3, h4, p, li").each((_index, element) => {
    const text = cleanText(getText($(element)));
    if (!text || isUnrelatedText(text) || seen.has(text)) {
      return;
    }

    seen.add(text);
    blocks.push(text);
  });

  if (blocks.length) {
    return blocks.join("\n");
  }

  const text = cleanText(getText(container));
  return isUnrelatedText(text) ? "" : text;
}

function isUnrelatedText(text) {
  const normalized = text.toLowerCase().replaceAll(" ", "");
  const socialFooter = "discord|x|facebook|instagram|tiktok|youtube|twitch";
  if (normalized === socialFooter) {
    return true;
  }

  const blocked = new Set([
    "download",
    "log in",
    "my account log out",
    "privacy policy",
    "term of use",
    "support",
    "comming soon",
  ]);
  return blocked.has(text.trim().toLowerCase());
}

function metaContent($, selector) {
  const tag = selectOne($, selector);
  if (tag === null) {
    return null;
  }

  const content = tag.attr("content");
  return content ? cleanText(String(content)) : null;
}

function normalizeArticleUrl(value, baseUrl) {
  const absoluteUrl = urldefrag(urljoin(baseUrl, value.trim()));
  const parsed = urlsplit(absoluteUrl);
  if ((parsed.scheme !== "http" && parsed.scheme !== "https") || !parsed.netloc) {
    return null;
  }

  return absoluteUrl;
}

function looksLikeHomepage(url) {
  const path = urlsplit(url).path.replace(/\/+$/, "");
  return path === "" || path === "/index.html";
}

function looksGenericTitle(value) {
  return value.toLowerCase().includes("super hero team-based pvp shooter");
}

function cleanTitle(value) {
  let title = cleanText(value);
  for (const separator of ["_Marvel Rivals", " | Marvel Rivals", " - Marvel Rivals"]) {
    if (title.includes(separator)) {
      title = title.split(separator)[0];
    }
  }
  return title.trim();
}

function cleanText(value) {
  return collapseWhitespace(String(value).replaceAll(" ", " "));
}

function truncateText(value, maxLength) {
  if (charLength(value) <= maxLength) {
    return value;
  }

  const head = sliceChars(value, 0, maxLength);
  const parts = rsplitOnce(head, "\n");
  const truncated = (parts.length === 2 ? parts[0] : parts[0]).trim();
  return truncated || head.trim();
}
